### PR TITLE
cmake: include config-specific and LTO flags into TARANTOOL_C_FLAGS

### LINKED ISSUE #8022: `tarantool.build.flags` doesn't include all compilation flags
**Steps to reproduce**

```lua
tarantool> require('tarantool').build.flags
---
- ' -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -Wformat -Wformat-security
  -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/git/tarantool=.
  -std=c11 -Wall -Wextra -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type
  -Werror'
...
```
or
```sh
$ tarantool -v
Tarantool 2.11.0-entrypoint-758-g85ef11180
Target: Linux-x86_64-Debug
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_BACKTRACE=FALSE
Compiler: GNU-8.3.0
C_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/git/tarantool=. -std=c11 -Wall -Wextra -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type -Werror
CXX_FLAGS: -fexceptions -funwind-tables -fno-common -fopenmp -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/git/tarantool=. -std=c++11 -Wall -Wextra -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type -Werror
```

**Actual behavior**

Note that build type is `Debug`, but there are neither `-g` nor `-O0` options.

**Expected behavior**

`TARANTOOL_C_FLAGS` should include `CMAKE_C_FLAGS_DEBUG` or `CMAKE_C_FLAGS_RELWITHDEBINFO`, depending on the build type. The same applies to `TARANTOOL_CXX_FLAGS`.

Also there might be CMake flags, e.g. `-DENABLE_LTO=true`, which passes `-flto` option to the compiler, however it is not added to `TARANTOOL_C_FLAGS` list, so it's hard to figure out from the final Tarantool binary whether it was compiled with LTO or not.

### PR DESCRIPTION
**Pull request by mkostoevr**
Prior to this patch only the flags given by the CMAKE_C_FLAGS were shown in the tarantool -v and Lua's tarantool.build.flags. Make it also show the flags from CMAKE_C_FLAGS_\<CONFIG\> and CFLAGS_LTO variables. The first one is config-specific set of flags (different for Debug, Release, RelWithDebInfo and MinSizeRel configs), the former is set in the cmake/lto.cmake if LTO is enabled.

Closes #8022

NO_DOC=build
NO_CHANGELOG=build

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -339,8 +339,24 @@ add_subdirectory(box)
 include_directories(${EXTRA_BOX_INCLUDE_DIRS})
 
 # Save CMAKE_XXX_FLAGS from this directory for config.h (used in --version)
-set(TARANTOOL_C_FLAGS ${CMAKE_C_FLAGS} PARENT_SCOPE)
-set(TARANTOOL_CXX_FLAGS ${CMAKE_CXX_FLAGS} PARENT_SCOPE)
+string (REPLACE ";" " " TARANTOOL_CFLAGS_LTO "${CFLAGS_LTO}")
```
Comment (Gumix) [inline src/CMakeLists.txt:342]: Nit: usually there are no space before `(`.

**Comment by Gumix** [comment]
Nit: Before this patch it was possible to build as `cmake .. -DCMAKE_BUILD_TYPE=None`, now:
```
CMake Error at src/CMakeLists.txt:348 (message):
  Unknown config passed: None
```
Dunno if it's ok or not.

**Review by Gumix — APPROVED** [review]
> `NO_TEST=build`

It's possible to create a test like [gh_7495_tarantool_build_target_test.lua](https://github.com/tarantool/tarantool/blob/master/test/app-luatest/gh_7495_tarantool_build_target_test.lua).

File: src/CMakeLists.txt (hunk 1 above)
Comment (mkostoevr) [reply src/CMakeLists.txt]: Done.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -339,8 +339,24 @@ add_subdirectory(box)
...
 # Save CMAKE_XXX_FLAGS from this directory for config.h (used in --version)
-set(TARANTOOL_C_FLAGS ${CMAKE_C_FLAGS} PARENT_SCOPE)
-set(TARANTOOL_CXX_FLAGS ${CMAKE_CXX_FLAGS} PARENT_SCOPE)
+string(REPLACE ";" " " TARANTOOL_CFLAGS_LTO "${CFLAGS_LTO}")
+
+set(TARANTOOL_BUILD_TYPES Debug Release RelWithDebInfo MinSizeRel)
+list(FIND TARANTOOL_BUILD_TYPES "${CMAKE_BUILD_TYPE}" TARANTOOL_BUILD_TYPE)
+
+if(${TARANTOOL_BUILD_TYPE} EQUAL -1)
+    message(FATAL_ERROR "Unknown config passed: ${CMAKE_BUILD_TYPE}")
```
Comment (mkostoevr) [inline src/CMakeLists.txt:348]: @Gumix, regarding this error: to me it's not good to silently accept misspelling a user has introduced, but I'm not sure if it's OK for Tarantool too. Awaiting for @locker's opinion.

**Review by locker — APPROVED** [review]

File: test/app-luatest/gh_8022_tarantool_build_flags_test.lua [hunk 3]
```diff
@@ -0,0 +1,10 @@
+local t = require('luatest')
+
+local g = t.group('gh-8022')
+
+-- Check that `tarantool.build.flags` contains an optimization level setting.
+-- It means that config-specific flags are included into the variable.
+g.test_build_target = function()
+    local tarantool = require('tarantool')
+    t.assert_str_contains(tarantool.build.flags, '-O')
```
Comment (locker) [inline test/app-luatest/gh_8022_tarantool_build_flags_test.lua:9]: This is a user-visible change. I think we need a changelog for it.

**Review by andreyaksenov — APPROVED** [review]

**Review by ochaplashkin — APPROVED** [review]

File: test/app-luatest/gh_8022_tarantool_build_flags_test.lua (hunk 3 above)
Comment (mkostoevr) [reply test/app-luatest/gh_8022_tarantool_build_flags_test.lua:9]: Done.

File: changelogs/unreleased/gh-8022-tarantool-build-flags.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+## feature/core
+
+* The `tarantool -v` output and `tarantool.build.flags` string now include
+  build type-specific and LTO flags if any.
```
Comment (locker) [inline changelogs/unreleased/gh-8022-tarantool-build-flags.md:4]:
```suggestion
  build type-specific and LTO flags if any (gh-8022).
```

File: changelogs/unreleased/gh-8022-tarantool-build-flags.md [hunk 5]
```diff
@@ -0,0 +1,4 @@
+## feature/core
```
Comment (locker) [inline changelogs/unreleased/gh-8022-tarantool-build-flags.md:1]:
```suggestion
## bugfix/core
```

Comment (mkostoevr) [reply changelogs/unreleased/gh-8022-tarantool-build-flags.md]: Done.

File: changelogs/unreleased/gh-8022-tarantool-build-flags.md (hunk 4 above)
Comment (mkostoevr) [reply changelogs/unreleased/gh-8022-tarantool-build-flags.md]: That was inaccurate, thanks.

**Review by locker — APPROVED** [review]

**Comment by Totktonada** [comment]
@mkostoevr It seems, it closes #3686, but, please, consider a comment regarding case sensetivity of `CMAKE_BUILD_TYPE`.

**Pull request merged**